### PR TITLE
[CoreImage] Update filters for Xcode 9

### PR DESCRIPTION
**Pull request by VincentDondain**
- Fixes bug #57350: Review new CoreImage filters added in Xcode 9
(https://bugzilla.xamarin.com/show_bug.cgi?id=57350).
- Adds `AVCameraCalibrationData` and `CIBarcodeDescriptor` to `generator-filters`.
- Fixes `ApiCoreImageFiltersTest`'s `GenerateBinding` to use valid `[CoreImageFilterProperty]`.
- In `CheckManagedFilters` generate code of SuperClass when detected so it's easier to bind.

### CODE REVIEW

**Comment by VincentDondain** [comment]
Adding `don't merge` because I have some questions (need to update the PR):

1. https://github.com/xamarin/xamarin-macios/pull/2974/files#diff-14b0758943a42f6b156f42dc8a2f8116R4838 how do we bind this?

2. https://github.com/xamarin/xamarin-macios/pull/2974/files#diff-14b0758943a42f6b156f42dc8a2f8116R4765, I see when turn on code generation that `CIMotionBlur.SuperClass == CILinearBlur` (`CILinearBlur == CIBlur`) but the availability attributes would be wrong for `CIBokehBlur`.

3. https://github.com/xamarin/xamarin-macios/pull/2974/files#diff-72e6c86015c96137c0f30e6be84aa8b4R122 is going to fail but I don't understand why. I left the code that puzzles me: https://github.com/xamarin/xamarin-macios/pull/2974/files#diff-72e6c86015c96137c0f30e6be84aa8b4R109 and https://github.com/xamarin/xamarin-macios/pull/2974/files#diff-72e6c86015c96137c0f30e6be84aa8b4R132, one works, the other doesn't.

File: src/coreimage.cs [hunk 1]
```diff
@@ -4729,70 +4733,76 @@ interface CIAttributedTextImageGenerator {
...
+		[CoreImageFilterProperty ("inputBarcodeDescriptor")]
+		CIBarcodeDescriptor BarcodeDescriptor { get; set; }
 	}
 
 	[CoreImageFilter]
 	[iOS (11,0)]
 	[Mac (10,13)]
 	[TV (11,0)]
-	[BaseType (typeof (CIFilter))]
+	[BaseType (typeof (CIFilter))] // Maybe 'CIScaleTransform' (shared 'Scale' and 'AspectRatio' property)?
```
Comment (spouliot) [inline src/coreimage.cs]: does `CIScaleTransform` exists ?
it's possible to add our own - but it can bite us back in the future if Apple introduce the same with different properties.

File: src/coreimage.cs [hunk 2]
```diff
@@ -4729,70 +4733,76 @@ interface CIAttributedTextImageGenerator {
...
+		float AspectRatio { get; set; }
 	}
 
 	[CoreImageFilter]
+	[Abstract]
 	[iOS (11,0)]
 	[Mac (10,13)]
 	[TV (11,0)]
 	[BaseType (typeof (CIFilter))]
+	interface CIBlur { // `CIMotionBlur.SuperClass == CILinearBlur` but it's marked as `[NoiOS]` and `[NoMac]` so we have ours
```
Comment (spouliot) [inline src/coreimage.cs]: same, does `CIBlur` exists ?

File: src/coreimage.cs [hunk 3]
```diff
@@ -4729,70 +4733,76 @@ interface CIAttributedTextImageGenerator {
...
+		[CoreImageFilterProperty ("inputRadius")]
+		float Radius { get; set; }
+	}
+
+	[CoreImageFilter]
+	[iOS (11,0)]
+	[Mac (10,13)]
+	[TV (11,0)]
+	[BaseType (typeof (CIBlur))] // `CIBokehBlur` doesn't explicitly have `SuperClass` but it should be `CIBlur` to inherit the `inputRadius` property
```
Comment (spouliot) [inline src/coreimage.cs]: same (about `CIBlur`)
note that adding a base type (in the future) would not be a breaking change if the flattened API is identical

File: src/coreimage.cs [hunk 4]
```diff
@@ -4816,67 +4827,77 @@ interface CIColorCurves {
...
-		//CIVector FocusRect { get; set; }
+		[CoreImageFilterProperty ("inputAperture")]
+		float Aperture { get; set; }
+
+		[CoreImageFilterProperty ("inputCalibrationData")]
+		AVCameraCalibrationData CalibrationData { get; set; }
+
+		// FIXME: How to bind this? Strongly typed?
+		// [CoreImageFilterProperty ("inputTuningParameters")]
+		// NSDictionary TuningParameters { get; set; }
```
Comment (spouliot) [inline src/coreimage.cs]: If we have documentation then let's do a strong dictionary.
If not then lets bind it as `WeakTuningParameters` so it's available/usable to developers

File: tests/introspection/ApiCoreImageFiltersTest.cs [hunk 5]
```diff
@@ -65,26 +65,6 @@ protected virtual bool Skip (string nativeName)
 			// Both reported in radar #21548819
 			//  NSUnknownKeyException [<CIDepthOfField 0x158586970> valueForUndefinedKey:]: this class is not key value coding-compliant for the key inputPoint2.
 			case "CIDepthOfField":
-			// NSUnknownKeyException [<CISunbeamsGenerator 0x1586d0810> valueForUndefinedKey:]: this class is not key value coding-compliant for the key inputCropAmount.
-			case "CISunbeamsGenerator":
```
Comment (spouliot) [inline tests/introspection/ApiCoreImageFiltersTest.cs]: ^ that one seems unrelated to xcode9 - please double check (git blame)

File: src/coreimage.cs (hunk 2 above)
Comment (spouliot) [reply src/coreimage.cs]: I do not see `CILinearBlur` anywhere in this PR or the existing sources

File: tests/monotouch-test/CoreImage/FilterTest.cs [hunk 6]
```diff
@@ -109,6 +117,47 @@ public void ColorSpace ()
...
+		public void CIBarcodeDescriptorTest ()
+		{
+			TestRuntime.AssertXcodeVersion (9, 0);
+
+			using (var f = new CIBarcodeGenerator ()) {
+				Assert.Null (f.BarcodeDescriptor, "CIBarcodeDescriptor/default");
+				using (var b = new CIQRCodeDescriptor ()) {
+					var nsname = NSString.CreateNative ("inputBarcodeDescriptor");
+					// TODO: remove when `CIBarcodeDescriptorTest` passes
+					// This doesn't work, `inputBarcodeDescriptor` is not set.
```
Comment (spouliot) [inline tests/monotouch-test/CoreImage/FilterTest.cs]: Unclear, you mean `f.BarcodeDescriptor == null` after `f.BarcodeDescriptor = b;` ?

The filter might have validation that requires other properties to be set before applying the generator ?

File: src/coreimage.cs (hunk 4 above)
Comment (VincentDondain) [reply src/coreimage.cs]: So the only documentation I know being https://developer.apple.com/library/content/documentation/GraphicsImaging/Reference/CoreImageFilterReference/index.html I guess the answer is no...

File: tests/introspection/ApiCoreImageFiltersTest.cs (hunk 5 above)
Comment (VincentDondain) [reply tests/introspection/ApiCoreImageFiltersTest.cs]: @spouliot removed because we do have this filter: https://github.com/xamarin/xamarin-macios/blob/master/src/coreimage.cs#L4563-L4594 (agree unrelated but that was easy to clean since I removed a bunch of cases.

**Comment by VincentDondain** [comment]
List of detected parents and their code: https://gist.github.com/VincentDondain/f4f1a4e171ea38fd61e66c7b020f8f5d

File: src/coreimage.cs (hunk 1 above)
Comment (VincentDondain) [reply src/coreimage.cs]: I attached generated code for the `SuperClass` in https://github.com/xamarin/xamarin-macios/pull/2974#issuecomment-343296118, `CIScaleTransform` isn't there so we'd have to create our own and yes that could potentially bite us which is why I added the comment :P

File: src/coreimage.cs (hunk 2 above)
Comment (VincentDondain) [reply src/coreimage.cs]: See https://github.com/xamarin/xamarin-macios/pull/2974#issuecomment-343296118, `CIBlur` doesn't exist but `CILinearBlur` does but it's marked as `[NoiOS]` and `[NoMac]` so I'm assuming it's not expected on iOS and Mac. That confuses me.

File: tests/monotouch-test/CoreImage/FilterTest.cs (hunk 6 above)
Comment (VincentDondain) [reply tests/monotouch-test/CoreImage/FilterTest.cs]: @spouliot I copied the `ColorSpace` test, there's a `f.ColorSpace = cs;` followed by a `f.ColorSpace = null;` there.

I'll dig into "the filter might have validation that requires other properties to be set", didn't think about that.

File: src/coreimage.cs (hunk 3 above)
Comment (spouliot) [reply src/coreimage.cs]: not if it's virtual

Comment (VincentDondain) [reply src/coreimage.cs]: If what is virtual? Nothing's virtual in `CIBokehBlur.g.cs` or `CIBlur.g.cs`

File: tests/monotouch-test/CoreImage/FilterTest.cs [hunk 7]
```diff
@@ -99,6 +103,10 @@ public void ColorSpace ()
 			using (var f = new CIColorCubeWithColorSpace ()) {
 				Assert.Null (f.ColorSpace, "ColorSpace/default");
 				using (var cs = CGColorSpace.CreateDeviceGray ()) {
+					var nsname = NSString.CreateNative ("inputColorSpace");
+					// TODO: remove when `CIBarcodeDescriptorTest` passes
+					// This works `inputColorSpace` is set
+					Messaging.void_objc_msgSend_IntPtr_IntPtr (f.Handle, Selector.GetHandle ("setValue:forKey:"), cs.Handle, nsname);
```
Comment (rolfbjarne) [inline tests/monotouch-test/CoreImage/FilterTest.cs]: This leaks `nsname`, you need to `NSString.ReleaseNative (nsname)`.

File: tests/monotouch-test/CoreImage/FilterTest.cs [hunk 8]
```diff
@@ -109,6 +117,47 @@ public void ColorSpace ()
...
+		{
+			TestRuntime.AssertXcodeVersion (9, 0);
+
+			using (var f = new CIBarcodeGenerator ()) {
+				Assert.Null (f.BarcodeDescriptor, "CIBarcodeDescriptor/default");
+				using (var b = new CIQRCodeDescriptor ()) {
+					var nsname = NSString.CreateNative ("inputBarcodeDescriptor");
+					// TODO: remove when `CIBarcodeDescriptorTest` passes
+					// This doesn't work, `inputBarcodeDescriptor` is not set.
+					Messaging.void_objc_msgSend_IntPtr_IntPtr (f.Handle, Selector.GetHandle ("setValue:forKey:"), b.Handle, nsname);
```
Comment (rolfbjarne) [inline tests/monotouch-test/CoreImage/FilterTest.cs]: Same: leaks `nsname`.

File: src/coreimage.cs (hunk 3 above)
Comment (spouliot) [reply src/coreimage.cs]: Allowed [1]
* _Introducing a new base class_
* _Moving a member onto a class higher in the hierarchy tree of the type from which it was removed_

[1] https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md

File: src/coreimage.cs (hunk 4 above)
Comment (spouliot) [reply src/coreimage.cs]: then a weak version should be exposed (it remain useful even if a strong dictionary is made available later)

File: src/coreimage.cs (hunk 2 above)
Comment (spouliot) [reply src/coreimage.cs]: it's been a while... and I remember mistakes in the values returned. However it could mean it's an _abstract_ type (so not _usable_, instead of not available, from iOS or Mac).

File: tests/monotouch-test/CoreImage/FilterTest.cs (hunk 7 above)
Comment (VincentDondain) [reply tests/monotouch-test/CoreImage/FilterTest.cs]: Yes yes but this code is going to be removed: `TODO: remove when 'CIBarcodeDescriptorTest' passes`.

It's just to illustrate that I can't get `setValue:forKey:` to work for `CIBarcodeGenerator` when it works fine for `CIColorCubeWithColorSpace`.

Comment (rolfbjarne) [reply tests/monotouch-test/CoreImage/FilterTest.cs]: Gotcha, clearly I just read what I want to read and ignore the rest..

**Comment by VincentDondain** [comment]
Removing `dont-merge` as the bullet points in https://github.com/xamarin/xamarin-macios/pull/2974#issuecomment-342871017 should all be addressed.

File: src/coreimage.cs [hunk 9]
```diff
@@ -4729,70 +4730,75 @@ interface CIAttributedTextImageGenerator {
...
-		//CIBarcodeDescriptor BarcodeDescriptor { get; set; }
+		[CoreImageFilterProperty ("inputBarcodeDescriptor")]
+		CIBarcodeDescriptor BarcodeDescriptor { get; set; }
 	}
 
 	[CoreImageFilter]
 	[iOS (11,0)]
 	[Mac (10,13)]
 	[TV (11,0)]
+	// Maybe 'typeof (CIFilter)' (shared 'Scale' and 'AspectRatio' property).
```
Comment (spouliot) [inline src/coreimage.cs]: not sure I understand the comment above ? the base type is already CIFilter

Comment (VincentDondain) [reply src/coreimage.cs]: Wrong copy and paste (:
